File: server-py/app/routers/sse.py
import asyncio
import json

from fastapi import APIRouter, Header, HTTPException
from starlette.responses import StreamingResponse

router = APIRouter(prefix="/api/sse", tags=["sse"])


@router.get("")
async def sse_stream(
    x_fsb_hash_key: str | None = Header(None, alias="X-FSB-Hash-Key"),
):
    """Server-Sent Events stream for real-time updates."""
    if not x_fsb_hash_key:
        raise HTTPException(status_code=401, detail="Missing X-FSB-Hash-Key header")

    # Import the shared client registry from agents router
    from .agents import sse_clients

    queue: asyncio.Queue = asyncio.Queue(maxsize=256)

    # Register client
    if x_fsb_hash_key not in sse_clients:
        sse_clients[x_fsb_hash_key] = []
    sse_clients[x_fsb_hash_key].append(queue)

    async def event_generator():
        try:
            # Send initial connected event
            yield f"data: {json.dumps({'type': 'connected'})}\n\n"

            while True:
                try:
                    # Wait for a message with a 30s timeout for keepalive
                    message = await asyncio.wait_for(queue.get(), timeout=30.0)
                    yield f"data: {message}\n\n"
                except asyncio.TimeoutError:
                    # Send keepalive comment
                    yield ": keepalive\n\n"
        except asyncio.CancelledError:
            pass
        finally:
            # Cleanup on disconnect
            clients = sse_clients.get(x_fsb_hash_key)
            if clients and queue in clients:
                clients.remove(queue)
                if not clients:
                    sse_clients.pop(x_fsb_hash_key, None)

    return StreamingResponse(
        event_generator(),
        media_type="text/event-stream",
        headers={
            "Cache-Control": "no-cache",
            "Connection": "keep-alive",
            "X-Accel-Buffering": "no",
        },
    )

File: server/src/routes/sse.js
const express = require('express');

function createSSERouter(sseClients) {
  const router = express.Router();

  // GET /api/sse - SSE stream for real-time updates
  router.get('/', (req, res) => {
    const hashKey = req.headers['x-fsb-hash-key'];
    if (!hashKey) {
      return res.status(401).json({ error: 'Missing X-FSB-Hash-Key header' });
    }

    // Set SSE headers
    res.writeHead(200, {
      'Content-Type': 'text/event-stream',
      'Cache-Control': 'no-cache',
      'Connection': 'keep-alive',
      'X-Accel-Buffering': 'no'
    });

    // Send initial connection event
    res.write(`data: ${JSON.stringify({ type: 'connected' })}\n\n`);

    // Register client
    if (!sseClients.has(hashKey)) {
      sseClients.set(hashKey, []);
    }
    sseClients.get(hashKey).push(res);

    // Send keepalive every 30 seconds
    const keepAlive = setInterval(() => {
      try {
        res.write(': keepalive\n\n');
      } catch {
        clearInterval(keepAlive);
      }
    }, 30000);

    // Cleanup on disconnect
    req.on('close', () => {
      clearInterval(keepAlive);
      const clients = sseClients.get(hashKey);
      if (clients) {
        const index = clients.indexOf(res);
        if (index !== -1) clients.splice(index, 1);
        if (clients.length === 0) sseClients.delete(hashKey);
      }
    });
  });

  return router;
}

module.exports = createSSERouter;
